### PR TITLE
feat(proofs): emit canonical Ethereum MPT-RLP from ProofNode

### LINKED ISSUE #2001: feat: Implement rust-side eth_getProof
**Is your feature request related to a problem? Please describe.**

Firewood currently exposes range and change proofs over its FFI, but there is no way to answer Ethereum's eth_getProof JSON-RPC against a Firewood-backed revision. Avalanche clients running on Firewood state can't serve eth_getProof requests, which blocks compatibility with off-the-shelf Ethereum tooling (wallets, explorers, light clients, indexers) that relies on the standard RPC for proving account and storage state.

**Describe the solution you'd like**

A new FFI call fwd_eth_get_proof(revision, account_key, slot_keys[]) that returns:

An array of canonical RLP-encoded MPT nodes forming the account proof.
For each requested slot, the slot value (or an absent marker) and an array of canonical RLP-encoded MPT nodes forming the storage proof.
A Go wrapper that turns those buffers into the standard eth_getProof JSON-RPC response shape. Inputs are pre-hashed 32-byte trie keys; the caller is responsible for keccak256(address) and keccak256(slot).

The native Firewood proof format is not RLP-encoded MPT, so an Ethereum verifier can't consume it directly. This feature emits the canonical wire shape verifiers expect.

**Describe alternatives you've considered**

Build eth_getProof in Go from the existing range-proof FFI. Rejected — the MPT RLP encoding (including the single-storage-slot synthetic root) is non-trivial and already exists in the Rust hasher. Duplicating it in Go is the wrong place for that knowledge.
Gate this code on the ethhash feature flag. Rejected — the ethhash flag is on track to be replaced by a runtime mode knob. Gating the new code on ethhash now would just produce #[cfg] walls to tear out later. (One narrow legacy-data exception is documented in the implementation plan.)
Return the JSON-RPC default-fields shape for absent accounts. Rejected at the FFI level in favor of a NotFound enum variant (matching how fwd_get_from_revision already signals "key absent" vs. real errors). The Go wrapper normalizes to the RPC shape when needed.

**Additional context**

Depends on PR #1938, which ensures account values written by a post-1938 binary carry the correct storageRoot field. Pre-1938 databases are still supported (the proof emitter recomputes on-the-fly for legacy data).
Best-effort behavior: the function will run against any trie. If the underlying trie was hashed with SHA-256 rather than Keccak-256, the emitted bytes are still structurally well-formed MPT RLP but won't verify under an Ethereum verifier — the caller is expected to know what database they opened.

### PR DESCRIPTION
**Pull request by rkuris**
## Why this should be merged

Step 2 of the `eth_getProof` plan. Adds the always-compiled encoder that turns Firewood `ProofNode`s into the canonical Ethereum Modified Merkle Patricia Trie wire format an `eth_getProof` verifier expects. Firewood's native proof layout is its own self-describing format; Ethereum tooling cannot consume it directly. This commit produces the buffers verifiers actually want, so a future Go FFI / JSON-RPC layer can serve `eth_getProof` against Firewood-backed state.

Stacked on #2003 (Step 1: lift `nibbles_to_eth_compact`).

## How this works

New module `firewood::proofs::eth` with two public functions:

- `proof_node_to_mpt_rlp(node) -> SmallVec<[Box<[u8]>; 2]>` emits one or two canonical MPT-RLP buffers. Two buffers are returned only when a single Firewood node carries both a non-empty partial path and a children-bearing branch whose RLP encoding is >= 32 bytes; in that case an MPT verifier needs both the inner branch (which it keccaks) and the outer extension (which references the inner by hash). At nibble depth 64 (32 bytes, the Keccak-256 of an account address) the value is treated as Ethereum account RLP and `storageRoot` is recomputed from the node's children.
- `synth_storage_leaf_rlp(storage_child)` builds the synthetic storage-trie leaf for the "account with exactly one storage slot" case. The plan named this parameter `account_node`, but constructing the synthetic leaf requires the storage child's partial path and value, which the account `ProofNode` does not carry; callers pass the storage child directly.

The bulk of the encoding mirrors the ethhash hasher (`storage/src/hashers/ethhash.rs::Preimage::write`). Where the hasher collapses encoded bytes into a hash, this module keeps the bytes so a verifier can walk them. A future refactor could share an `RlpEmit` trait between the two; deliberately deferred to keep this PR focused.

### Side-effects: nudging toward "ethhash as a runtime knob"

This PR takes a few opportunistic steps toward making the eventual flip from compile-time feature to runtime mode a smaller change:

- **Ungated `firewood_storage::fix_account_storage_root_value`.** The function is now unconditionally compiled. The variant-specific pattern-matching (`HashType::Hash` / `HashType::Rlp`) is factored into two small `#[cfg]`-gated helpers (`single_child_storage_root` and `child_to_rlp_item`). `sha3` moves from optional to unconditional in storage's deps.
- **`ValueDigest::value() -> Option<&[u8]>`** absorbs the variant match for "give me the value bytes," letting callers avoid `#[cfg(not(feature = "ethhash"))]` arms.
- **`cfg!(feature = "ethhash")` over `#[cfg]`** where the code shape is identical in both modes (notably the `is_account` boolean). The compiler folds the const to false in non-ethhash mode and LTO removes the dead-code call to `fix_account_storage_root_value`.
- `storage::rlp::{encode_list, RlpItem, replace_list_field}` were previously re-exported only under `cfg(any(test, feature = "test_utils"))`. Now unconditionally `pub` so the always-compiled emitter can use them.

The remaining `#[cfg(feature = "ethhash")]` gates in the touched files are the ones where the body literally pattern-matches enum variants that only exist under ethhash, and therefore cannot fold via `cfg!`.

## How this was tested

- Five new unit tests covering leaf encoding, empty-value handling, and `synth_storage_leaf_rlp` boundary cases. All pass under both default features and `--features ethhash,logger`.
- Full workspace test suite passes under both configurations.
- `cargo clippy --workspace --features ethhash,logger --all-targets` clean.
- `cargo doc --no-deps --features ethhash,logger` clean.
- `cargo fmt --check` clean.

## Breaking Changes

- `firewood-storage`: `fix_account_storage_root_value` is no longer gated on `ethhash`; it's always available. Previously it was only callable when the `ethhash` feature was enabled. No callers outside the workspace are known to be affected.

### CODE REVIEW

File: storage/src/hashednode.rs [hunk 1]
```diff
@@ -109,6 +109,16 @@ impl<T: AsRef<[u8]>> ValueDigest<T> {
...
 
+    /// Returns the inner bytes if this digest carries a value, or `None` if
+    /// it carries only a hash (`#[cfg(not(feature = "ethhash"))]` only).
+    pub fn value(&self) -> Option<&[u8]> {
+        match self {
+            Self::Value(v) => Some(v.as_ref()),
+            #[cfg(not(feature = "ethhash"))]
+            Self::Hash(_) => None,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (rkuris) [reply storage/src/hashednode.rs:121]: Fixed in 76d8dbe764de.

File: firewood/src/proofs/eth.rs [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+    // The account-node concept only exists in ethhash mode. Without that
+    // feature, depth-64 nodes are plain MPT nodes and no special handling
+    // fires.
+    let is_account = cfg!(feature = "ethhash") && key.len() == ACCOUNT_DEPTH_NIBBLES;
+    let value_bytes = node.value_digest.as_ref().and_then(ValueDigest::value);
+    let child_count = node
+        .child_hashes
+        .iter()
+        .filter(|(_, c)| c.is_some())
+        .count();
```
[the remark this answers is not in the document]
Comment (rkuris) [reply firewood/src/proofs/eth.rs]: Fixed in 76d8dbe764de. Also folded the count into the `if` since the temporary added no clarity.

File: firewood/src/proofs/eth.rs [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+        // (which the verifier keccaks to confirm the reference).
+        let hash = Keccak256::digest(&inner_bytes);
+        let outer = encode_list(&[
+            RlpItem::Bytes(&compact_path),
+            RlpItem::Bytes(hash.as_slice()),
+        ]);
+        let mut out: SmallVec<[Box<[u8]>; 2]> = SmallVec::new();
+        out.push(inner_bytes);
+        out.push(outer);
+        out
```
[the remark this answers is not in the document]
Comment (rkuris) [reply firewood/src/proofs/eth.rs:148]: Fixed in 76d8dbe764de. Flipped to `[outer, inner]` so the buffers come back in proof walk order; callers can concat directly into `accountProof[]`. Doc and the corresponding test updated to match.

File: firewood/src/proofs/eth.rs [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+
+#[cfg(not(feature = "ethhash"))]
+fn child_to_rlp_item(child: Option<&HashType>) -> RlpItem<'_> {
+    // Without ethhash, HashType is just a 32-byte TrieHash and there is no
+    // inline-RLP form; every present child is referenced by hash.
+    match child {
+        Some(hash) => RlpItem::Bytes(hash.as_slice()),
+        None => RlpItem::Empty,
+    }
+}
```
[the remark this answers is not in the document]
Comment (rkuris) [reply firewood/src/proofs/eth.rs:227]: Fixed in 76d8dbe764de. Renamed the eth.rs helper to `proof_child_rlp_item` and added a doc note pointing at the storage version and explaining why this one surfaces `HashType::Rlp` as `RlpItem::Raw` instead of `unreachable!`.

File: firewood/src/proofs/eth.rs [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+    // Compute the storage trie's root from this node's children and splice
+    // it into the account RLP value at field index 2.
+    let inner_bytes: Box<[u8]> = if is_account {
+        match value_bytes.and_then(|v| fix_account_storage_root_value(v, &node.child_hashes)) {
+            Some(spliced) => spliced,
+            None => branch_bytes,
+        }
+    } else {
+        branch_bytes
+    };
```
[the remark this answers is not in the document]
Comment (rkuris) [reply firewood/src/proofs/eth.rs:127]: Fixed in 76d8dbe764de. The fallback now mirrors the hasher (`ethhash.rs:234-236`): when the splice fails on a node that does carry a value, emit the raw account RLP instead of the value-less branch. The `None` value_bytes case still falls through to `branch_bytes`, which is correct (nothing to splice).

File: firewood/src/proofs/eth.rs [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+/// node the extension references by hash. The two buffers are returned in
+/// **proof walk order** `[outer_extension, inner_branch]` so callers can
+/// concatenate the result into a root-to-leaf proof list without
+/// reordering. The inline-extension case (branch RLP < 32 bytes, or any
+/// account node) collapses into a single self-contained buffer.
+#[must_use]
+pub fn proof_node_to_mpt_rlp(node: &ProofNode) -> SmallVec<[Box<[u8]>; 2]> {
+    // `partial_len` is validated against `key.len()` during deserialization,
+    // so the slice from `partial_len..` is always in range; the `unwrap_or`
+    // is just to satisfy clippy's panic-on-slice lint.
```
Comment (demosdemon) [inline firewood/src/proofs/eth.rs]: nit: useful to mention the lint name literally for easy searching

```suggestion
    // is just to satisfy clippy::indexing_slicing
```

**Review by demosdemon — APPROVED** [review]

File: firewood/src/proofs/eth.rs (hunk 6 above)
Comment (rkuris) [reply firewood/src/proofs/eth.rs]: Good point. Fixed in 2239d9d6629f.
